feat(widget): pretty printing restricted quantifiers

EdAyers opened the pull request:
Eg converts `∀ (x : ℕ), x > 3 → ∃ (N : set ℕ), ∀ y, y ∈ N → x > 2`
to `∀ (x > 3), ∃ (N : set ℕ), ∀ (y ∈ N), x > 2`
in the widget view.

![image](https://user-images.githubusercontent.com/5064353/102696261-f2852500-4224-11eb-807b-41e7ce9efdbe.png)

---

- [x] depends on: #5895
- [ ] add an option to disable
- [ ] automatically disable if `pp.notation` is off

EdAyers committed:
feat(widget): pretty printing restricted quantifiers

Eg converts `∀ (x : ℕ), x > 3 → ∃ (N : set ℕ), ∀ y, y ∈ N → x > 2`
to `∀ (x > 3), ∃ (N : set ℕ), ∀ (y ∈ N), x > 2`
in the widget view.

Things to do:
- testing
- include support for other binders, not just ∀
- docstrings

EdAyers committed:
Fix dodgy addressing

b-mehta commented on src/tactic/interactive_expr.lean:
```diff
@@ -65,6 +67,52 @@ meta def sf.of_eformat : eformat → sf
+
+meta def sf.collapse_restricted_quantifiers_pred : expr → tactic bool
+| `(@gt _ _ %%(expr.var 0) _) := pure tt
+| `(@has_mem.mem _ _ _ %%(expr.var 0) _) := pure tt
```
Perhaps I'm missing something, but what about `x < 2`? Do you also want a case here for `lt` and `le`?

EdAyers replied:
:)

EdAyers replied:
It's an mwe rn

EdAyers committed:
Add docs

EdAyers committed:
Add more preds

EdAyers committed:
Remove breaking example

EdAyers committed:
Merge branch 'master' into collapse-restricted-quantifiers

EdAyers committed:
add rudimenrary support for ∃.

I think there might be a bug in the  pretty printer.

EdAyers committed:
Clean up

EdAyers committed:
Merge branch 'master' into collapse-restricted-quantifiers

EdAyers committed:
Fix issue where nested quantifiers stopped collapsing

EdAyers committed:
Remove example

robertylewis commented on src/tactic/interactive_expr.lean:
```diff
@@ -65,6 +67,80 @@ meta def sf.of_eformat : eformat → sf
 | (compose x y) := sf.compose (sf.of_eformat x) (sf.of_eformat y)
 
+/-- `subtract x y` is some `z` when `∃ z, x = y ++ z` [fixme] this probably already exists. -/
+def subtract {α} [decidable_eq α]: (list α) → (list α) → option (list α)
```
This is `list.get_rest`.

`#find` is a convenient way to find things like this:
```lean
import data.list.defs tactic.find

#find list _ → list _ → option (list _)
```

robertylewis commented on src/tactic/interactive_expr.lean:
```diff
@@ -65,6 +67,80 @@ meta def sf.of_eformat : eformat → sf
+| l (sf.compose a b) := sf.follow l a <|> sf.follow l b
+
+/-- Run `f` on each of the immediate child substrings of the given string. -/
+meta def sf.traverse {m} [applicative m] (f : sf → m sf) : sf → m sf
```
There's a `traversable` type class, is that relevant here?

EdAyers replied:
Yeah it's not quite the same thing as the typeclass because the class's arg is a functor. I've seen traverse used for both of these though

robertylewis commented on src/tactic/interactive_expr.lean:
```diff
@@ -65,6 +67,80 @@ meta def sf.of_eformat : eformat → sf
+meta def sf.replace {m} [monad m] [alternative m] (f : sf → m sf) : sf → m sf
+| x := (f x >>= sf.traverse sf.replace) <|> pure x
+
+/-- The test for whether the proposition is a valid target for restricted quantifier collapsing.
```
Aren't the valid targets "everything with infix notation"?

EdAyers replied:
Is there a programmatic way to test whether a given relation has an infix notation?

robertylewis replied:
Not that I'm aware of. You could build a dummy application and see if it pretty prints differently than you expect, which is very hackish.

robertylewis commented on src/tactic/interactive_expr.lean:
```diff
@@ -65,6 +67,80 @@ meta def sf.of_eformat : eformat → sf
+  expr.lam _ _ py pred ← pure pred,
+  should_collapse ← sf.collapse_restricted_quantifiers_pred py,
+  if ¬ should_collapse then pure s else do
+  a1 ← pure $ [app_arg, lam_var_type], -- [hack] There is a bug in the addresing code. Second binder has the address of the first binder!
```
Is this bug reported somewhere more visible?

EdAyers replied:
Fixed now in https://github.com/leanprover-community/lean/pull/520

EdAyers replied:
This code will need to be changed once the fix has made it downstream

bryangingechen replied:
I've marked this dependent on #5895, which migrates mathlib to 3.26.0c which includes the fix.

EdAyers committed:
Merge branch 'master' into collapse-restricted-quantifiers

EdAyers committed:
Remove list.subtract

EdAyers committed:
Add note regarding addressing bug

EdAyers committed:
Add collapsed quantifiers to clipboard paste code.

For @PatrickMassot 👁‿👁
I do this by refactoring the sf handling code in to its own method that
returns the sf tree with the new look.

EdAyers committed:
Merge branch 'master' into collapse-restricted-quantifiers

EdAyers committed:
Fix change for #5895

EdAyers committed:
Oops, premature commit
